### PR TITLE
[SG2] Preview System Restructure and Time Node support

### PR DESCRIPTION
**Pull request by Nightmask3**
### Purpose of this PR

Initial setup for work needed to rearchitect preview system for better maintainability and performance.

Adds animated time node support for previews.

This PR also removes any references to the preview manager. (except in the test code)

https://user-images.githubusercontent.com/13033542/186270545-15d88f89-197c-4cc8-8702-cbfbdbe57356.mp4

---
### Testing status

Known issues:

1) Saving resets node previews to their default texture and they dont display correct output on loading from a saved graph (this is a regression that is my highest priority to fix)
2) Not async-compile enabled yet.
3) Time node support might be spotty for some nodes.
4) This breaks all our preview manager related tests, but will fix those soon.
5) Some preview manager interactions are still not operational, but the preview manager output itself should work fine.

This branch is not super well tested, but any discovered issues will be addressed in a quick followup PR.

---
### Comments to reviewers
Notes for the reviewers you have assigned.

### CODE REVIEW

File: com.unity.graphtools.foundation/Editor/UI/Views/ModelInspector/InspectorSelectionObserver.cs [hunk 1]
```diff
@@ -44,9 +46,19 @@ public override void Observe()
                             selectedModels.Add(graphModel);
                         }
 
+                        // GTF-EDIT: Caching the view state if the selection hasnt changed
+                        if (m_LastSelection?.Count() == 1 && selectedModels.Count() == 1)
+                        {
+                            var sgModel1 = selectedModels.FirstOrDefault();
```
Comment (Nightmask3) [inline com.unity.graphtools.foundation/Editor/UI/Views/ModelInspector/InspectorSelectionObserver.cs:52]: TODO: Split off into another PR

File: com.unity.graphtools.foundation/Editor/UI/Views/GraphView/GraphView.cs [hunk 2]
```diff
@@ -1376,7 +1376,9 @@ protected void OnDoubleClick()
         {
             // Display graph in inspector when clicking on background
             // TODO: displayed on double click ATM as this method overrides the Token.Select() which does not stop propagation
-            Selection.activeObject = GraphViewModel.GraphModelState?.GraphModel.Asset as Object;
+
+            // GTF-EDIT: Do not want this behavior, we should ask them to make it optional
+            //Selection.activeObject = GraphViewModel.GraphModelState?.GraphModel.Asset as Object;
```
Comment (Nightmask3) [inline com.unity.graphtools.foundation/Editor/UI/Views/GraphView/GraphView.cs:1381]: TODO: Split off into a separate PR

Comment (Nightmask3) [reply com.unity.graphtools.foundation/Editor/UI/Views/GraphView/GraphView.cs:1381]: And also ask GTF what their intention is around having this behavior

File: com.unity.sg2/Editor/GraphDeltaRegistry/PreviewManager/IPreviewUpdateReceiver.cs [hunk 3]
```diff
@@ -0,0 +1,9 @@
+﻿using UnityEngine;
+
+namespace UnityEditor.ShaderGraph.GraphDelta
+{
+    public interface IPreviewUpdateReceiver
```
Comment (Nightmask3) [inline com.unity.sg2/Editor/GraphDeltaRegistry/PreviewManager/IPreviewUpdateReceiver.cs:5]: Rename to `IPreviewDataStore` or something less conflicting with 'IPreviewUpdateListener'

File: com.unity.sg2/Editor/GraphDeltaRegistry/GraphDelta/GraphDelta.cs [hunk 4]
```diff
@@ -193,6 +193,10 @@ public NodeHandler GetNode(ElementID id, Registry registry)
         public void RemoveNode(ElementID id, Registry registry)
         {
             var node = GetNode(id, registry);
+
+            if (IsTimeDependentNode(node))
```
Comment (Nightmask3) [inline com.unity.sg2/Editor/GraphDeltaRegistry/GraphDelta/GraphDelta.cs:197]: TODO: Remove from here and do this at a UI-level instead

File: com.unity.sg2/Editor/GraphUI/EditorCommon/Preview/PreviewUpdateDispatcher.cs [hunk 5]
```diff
@@ -0,0 +1,168 @@
...
+                m_LastTimedUpdateTime = curTime;
+            return update;
+        }
+
+        public void Update()
+        {
+            if (m_OwningWindowReference == null)
+                return;
+
+            if (TimedNodesShouldUpdate(m_OwningWindowReference))
```
Comment (Nightmask3) [inline com.unity.sg2/Editor/GraphUI/EditorCommon/Preview/PreviewUpdateDispatcher.cs:163]: TODO: Do this in the PreviewService side of things instead, where we just need to update time in MPB and service can figure out halo and who to update

File: com.unity.sg2/Editor/GraphUI/EditorCommon/Preview/PreviewUpdateDispatcher.cs [hunk 6]
```diff
@@ -0,0 +1,168 @@
...
+                m_TimeDependentNodes.Add(listenerID);
+
+            // If a node was deleted we don't want to issue update calls for it thereafter
+            if (wasNodeDeleted)
+            {
+                impactedNodes.Remove(listenerID);
+                m_TimeDependentNodes.Remove(listenerID);
+            }
+
+            foreach (var downstreamNode in impactedNodes)
```
Comment (Nightmask3) [inline com.unity.sg2/Editor/GraphUI/EditorCommon/Preview/PreviewUpdateDispatcher.cs:99]: Can we move evaluation of halo into the preview service? so that update dispatcher isnt responsible

**Review by esmelusina — APPROVED** [review]
Looks good and matches the design-- thanks for taking those few things out!

**Review by brett-groshong — APPROVED** [review]
